### PR TITLE
Allow a callback for Session[Initializing|Closing]Async

### LINKED ISSUE #29917: [FEATURE REQ] Allow a callback for SessionInitializingAsync for SB functions
### Library name

 Microsoft.Azure.WebJobs.Extensions.ServiceBus 5.5.1

### Please describe the feature.

The extension already allows a call back for [`ProcessErrorAsync` in the options](https://github.com/Azure/azure-sdk-for-net/blob/19b4e800e3711e6cb0c925d60df4a4aa957eed92/sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs#L131)  it would be helpful to also expose `SessionInitializingAsync`, and `SessionClosingAsync` using the same pattern.

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes https://github.com/Azure/azure-sdk-for-net/issues/29917

- [x] Approve API changes
- [x] Update Change Log

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Review by jsquire — COMMENTED** [review]
Thanks, @danielmarbach!    I'm going to defer to @JoshLove-msft for the authoritative review, but this looks good to me!

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs [hunk 1]
```diff
@@ -206,6 +224,22 @@ internal async Task ExceptionReceivedHandler(ProcessErrorEventArgs args)
             }
         }
 
+        internal async Task SessionInitializingHandler(ProcessSessionEventArgs args)
+        {
+            if (SessionInitializingAsync != null)
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs]: Done

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs [hunk 2]
```diff
@@ -130,6 +130,24 @@ public int MaxConcurrentSessions
         /// </summary>
         public Func<ProcessErrorEventArgs, Task> ProcessErrorAsync { get; set; }
 
+        /// <summary>
+        /// Optional handler that can be set to be notified when a new session is about to be processed.
+        /// </summary>
+        /// <remarks>
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Config/ServiceBusOptions.cs]: Removed

**Comment by danielmarbach** [comment]
Where do I need to approve the API changes? For some reasons, I get no local build failure.

**Comment by jsquire** [comment]
> Where do I need to approve the API changes? For some reasons, I get no local build failure.

CI is just complaining that `<ROOT>/eng/scripts/Export-API.ps1 -ServiceDirectory "servicebus/Azure.Messaging.ServiceBus"` needs to be run.

**Comment by danielmarbach** [comment]
Isn't ApiDiff built into dotnet build in the main lib? Why does this one require a dedicated powershell script?

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusSessionsEndToEndTests.cs [hunk 3]
```diff
@@ -574,6 +574,46 @@ public async Task TestSingle_ReceiveFromFunction()
...
+            var host = BuildHost<TestCustomSessionClosingHandler>(SetCustomSessionInitializingHandler);
+            using (host)
+            {
+                bool result = _waitHandle1.WaitOne(SBTimeoutMills);
+                Assert.True(result);
+                await host.StopAsync();
+            }
+        }
+
+        protected static Action<IHostBuilder> SetCustomSessionClosingHandler =>
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusSessionsEndToEndTests.cs:610]: nit: this can be private I believe.

**Review by JoshLove-msft — APPROVED** [review]
Looks awesome!

**Comment by JoshLove-msft** [comment]
Looks like the new tests are failing CI. The error seems to indicate that there was an error that was logged as part of the test, but unfortunately doesn't report the error message. I can try to dig into this tomorrow if you don't have time @danielmarbach.

**Comment by danielmarbach** [comment]
@JoshLove-msft might be related to the thing I fixed (see my [awesome commit message](https://github.com/Azure/azure-sdk-for-net/pull/29945/commits/6e0c06181077f8da0f3f7f6d502b360c94c62b36))

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft might be related to the thing I fixed (see my [awesome commit message](https://github.com/Azure/azure-sdk-for-net/pull/29945/commits/034745473c582c5b338b6bb4f3d68947fa13a405))

Hah, nice!

**Comment by danielmarbach** [comment]
@JoshLove-msft it looks like the session initializing was not triggering. I added proper usage of the session. Hopefully, that resolve it. Or do you need to explicitly send with multiple session ids to trigger it?

**Comment by JoshLove-msft** [comment]
> @JoshLove-msft it looks like the session initializing was not triggering. I added proper usage of the session. Hopefully, that resolve it. Or do you need to explicitly send with multiple session ids to trigger it?

You shouldn't need to send multiple sessions. Let me kick off the tests again.

**Comment by danielmarbach** [comment]
Ok I need to properly setup the local environment again to debug this. I have paved my machine recently but I guess it is time to go through setting up things because the failing tests give me no insights

**Comment by jsquire** [comment]
> On my linux box. Will need to figure out what's going on at some point

Have you run the script directly and captured the environment variables?    It seems that we may need to look at the "run automatically" code in the test framework and make sure it's not bound to the SDK environment.    I know that we validated the script itself, though.

File: sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusSessionsEndToEndTests.cs [hunk 4]
```diff
@@ -1106,6 +1146,41 @@ public class TestReceiveFromFunction
...
+                [ServiceBusTrigger(FirstQueueNameKey, IsSessionsEnabled = true)]
+                ServiceBusReceivedMessage message,
+                ServiceBusSessionMessageActions sessionActions)
+            {
+                await sessionActions.CompleteMessageAsync(message);
+            }
+
+            public static Task SessionInitializingHandler(ProcessSessionEventArgs arg)
+            {
+                _waitHandle1.Set();
```
Comment (JoshLove-msft) [inline sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusSessionsEndToEndTests.cs:1161]: I think the issue is that this effectively ends the test before the message is completed, and we have validation that runs at the end of each test to ensure all messages have been drained from the queue. I think we can probably just combine these two tests into one test and set the waitHandle in the closing handler. We would need to set some other private field from the initializing handler, that we can check in the function body.

Comment (danielmarbach) [reply sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/tests/ServiceBusSessionsEndToEndTests.cs:1146]: Turned it into one test now.

**Comment by danielmarbach** [comment]
Yehaaaa!

**Pull request merged**